arch: Added initial OpenRISC port

jhol opened the pull request:
This PR follows on from #83933 which was closed due to inactivity. It has been many months since I have had any time to put into up-streaming these patches, so the PR was automatically closed.

Here is v9 of the patch-set with the following changes:

  - Rebased on Zephyr v4.2.1
      - Update to picolibc version is already upstream.
  - Support non-multi-threaded operation.
  - Fix stack initialization (thanks @NomanKhan777).
  - Fix typo `Interatively`.
  - Don't use `z_openrisc_thread_start` for `SECTION_FUNC`, because we don't want a section division here when `_isr_tail` falls through to it.
  - Ensure `r0` is initialized to zero (thanks @zgliu799).
  - Moved interrupt control functions into `irq.h` and make them inline.
  - Support multi-level IRQ handling.
  - Moved `__reset` and `__start` labels into reset exception handler.
  - Removed redundant call to `z_check_stack_sentinel` in `_isr_tail`. It is already called in `z_get_next_switch_handle`.
  - Call `z_check_stack_sentinel` in `z_openrisc_handle_irqs`.
  - Save the usage of `r15` in `_isr_tail`.
  - `load_caller_saved` afer `load_mac` and `load_esr_epcr` for symmetry with storage macros.
  - Various linker script improvements:
      - Add support for XIP.
      - Place snippets before `noload` and `bss` sections so as not to pad output image.
      - Support device-tree regions.
  - Added heading to `include/zephyr/arch/openrisc/arch.h` similar to match `include/zephyr/arch/mips/arch.h`.

The Zephyr port for the proprietary cousin of this architecture has now entered full production.

@zgliu799 has tested v8 version of the patches running an OpenRISC processor on an FPGA, and reports that they have worked successfully. He has also test the build using the Zephyr or1k tool-chain from the [sdk-ng PR](https://github.com/zephyrproject-rtos/sdk-ng/pull/864).

~~The sdk-ng patches are the main roadblock to getting this PR merged. They have been tested and rebased on the sdk-ng `main` branch.~~

jhol committed:
toolchain: common: Define ALIGN and PERFOPT_ALIGN for OpenRISC

.balign is available in the or1k GNU assembler, therefore ALIGN and
PERFOPT_ALIGN are both configured use it.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

FIX: balign

jhol committed:
toolchain: gcc: Added symbol attributes for OpenRISC

A common set of macros: GTEXT(), GDATA(), WTEXT() and WDATA() is defined
for Arm, Arm66, MIPS, RISC-V, RX and Xtensa. This patch adds OpenRISC
to this set.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Define output format for or1k-elf

The OpenRISC or1k-elf GCC compiler output format is named "elf32-or1k".
This patch adds an OUTPUT_FORMAT linker macro for this platform.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
soc: qemu: Added qemu_or1k simulated SoC

Qemu includes support for the OpenRISC 1000 CPU architecture. This patch
adds a Zephyr SoC definition which enables usage of this feature.

The SoC definition closely mirrors the Qemu MIPS Malta SoC definition.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
cmake: emu: qemu: Add binary suffix for or1k architecture

The Qemu emulator executable for the OpenRISC 1000 (or1k) architecture is
named qemu-or1k. This patch adds this information to the qemu architecture
suffix list.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
debug: thread_info: Store thread stack pointer offset for OpenRISC

In OpenRISC 1000, the stack pointer is stored in the r1 register. This
patch stores the offset of this value within in the thread structure into
the thread info offsets.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
scripts: logging: Add openrisc to ARCHS

The OpenRISC CPU architecture is signified by the CONFIG_OPENRISC Kconfig
option. This patch adds this information to the log database ARCHS table.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
scripts: twister: Add OpenRISC as a supported test platform

Adds the openrisc CPU architecture to the twister platform definitions.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
toolchain: gcc: Add GEN_ABSOLUTE_SYM definitions for OpenRISC

Existing versions of GEN_ABSOLUTE_SYM and GEN_ABSOLUTE_SYM_KCONFIG are
defined for all supported architectures. This patch adds a definition of
the same kind used for MIPS, RISC-V and Xtensa.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
openrisc: linker: Use zephyr's convention for rodata

Zephyr maps the start and end of the rodata section with variables using
the __rodata_region_start and __rodata_region_end. This patch adopts this
convention for the OpenRISC architecture.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
tests: Add basic test-set for OpenRISC architecture port

The OpenRISC 1000 architecture can be tested through the qemu_or1k emulated
SoC and board. Because this emulated device has minimal external hardware
by default, this patch enables a minimal suite of tests covering core
kernel features for the CPU architecture.

When running the test suite, OpenRISC was found to require additional stack
space to prevent an overflow. Therefore, a minimal additional amount of
storage: 128-bytes, was added that was found in practice to allow tests to
complete.

Note that Qemu version 9.2.0 or newer is required so as to include the
following commit:

    commit 3eb43aeb164f1f83c97ff693c7d464b49755110c
    Author: Joel Holdsworth <jholdsworth@nvidia.com>
    Date:   Fri Jun 7 15:29:33 2024 -0700

        hw/openrisc: Fixed undercounting of TTCR in continuous mode

This fixes a bug in the implementation of the emulated OpenRISC Tick Timer
which prevents the Zephyr OpenRISC port from task-scheduling properly.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
arch: Added initial OpenRISC architecture port

This patch adds support for the OpenRISC 1000 (or1k) architecture: a
MIPS-like open hardware ISA which was first introduced in 2000.

The thread switching implementation uses the modern Zephyr thread "switch"
architecture.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
drivers: timer: Add timer driver for or1k

The OpenRISC 1000 Tick Timer is tightly coupled to the or1k CPU core, and
is explicitly designed to facilitate task scheduling and high-resolution
timing.

The timer is documented in Chapter 14 of the OpenRISC 1000 Architecture
Manual:

    https://openrisc.io/or1k.html#__RefHeading__504849_595890882

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
boards: qemu: add board definition for the or1k architecture

Qemu includes support for the OpenRISC 1000 CPU architecture. This patch
adds a Zephyr virtual board definition which enables usage of this feature.

The board definition closely mirrors the Qemu MIPS Malta board definition.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

nordicjm commented:
Can you explain why openrisc is a separate architecture to risc (I don't have knowledge on them but assumed they would be the ISA compatible?)

maass-hamburg commented:
> Can you explain why openrisc is a separate architecture to risc (I don't have knowledge on them but assumed they would be the ISA compatible?)

did you meant RISC**V**? they are completely different, just like ARM and RISCV are

maass-hamburg commented:
> The Zephyr CI system is not yet using the new tool-chain. Is it possible to update it for the new architecture?

you have to wait until the next sdk version is released ( see https://github.com/zephyrproject-rtos/sdk-ng/issues/933) and also the reference changed to the new version in zephyr like in https://github.com/zephyrproject-rtos/zephyr/pull/95344

nashif commented:
confirming that the next SDK release will have the openrisc toolchain integrated, so that should be soon...

jhol commented:
v8:

  * Added missing include for `arch_bss_zero`

maass-hamburg reviewed (dismissed):
LGTM, but the ci errors need fixing

keith-packard committed:
arch/openrisc: Add THREAD_LOCAL_STORAGE support

Clear the TLS base pointer (r10) in arch_kernel_init.
Allocate the TLS area in arch_tls_stack_setup.
Set the TLS base pointer register (r10) in arch_new_thread.
Set ARCH_HAS_THREAD_LOCAL_STORAGE for config OPENRISC.

Signed-off-by: Keith Packard <keithp@keithp.com>

kartben commented on boards/qemu/or1k/doc/index.rst:
```diff
@@ -0,0 +1,92 @@
+
+OpenRISC 1000 Emulation (QEMU)
+##############################
+
```
```suggestion
```

kartben commented on boards/qemu/or1k/doc/index.rst:
```diff
@@ -0,0 +1,92 @@
+------------
+
+Qemu Tick Timer timer uses a clock frequency of 20 MHz,
+see hw/openrisc/cputimer.c in Qemu source tree for details.
```
```suggestion
see ``hw/openrisc/cputimer.c`` in Qemu source tree for details.
```

kartben reviewed (commented):
once again doc clean up can be a follow-up, no worries at all :)

nashif committed:
qemu_or1k: fix uart register in DTS

UART has the wrong register. Used to work with older Qemu, not in the
10 release we have in the Zephyr SDK.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
libc: use picolibc module with openrisc [REVERT ME]

Temporary fix while we wait for new SDK release.

See https://github.com/zephyrproject-rtos/sdk-ng/pull/1106

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
qemu_or1k: limit tests to kernel and arch initially [REVERT ME]

Initially, limit tests to only kernel and architecture. Should be
reverted once we fix issues in the Zephyr SDK.

See https://github.com/zephyrproject-rtos/sdk-ng/pull/1106

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: openrisc: do not enableL TLS support [REVERT ME]

Disable TLS while we wait for toolchain update in the new Zephyr SDK.

See https://github.com/zephyrproject-rtos/sdk-ng/pull/1106

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
